### PR TITLE
Implement beta support for discovering mulled containers for dependencies.

### PR DESCRIPTION
**Pull request by jmchilton**
_tl;dr - Free community managed Docker containers for all tools whose requirements are resolvable with bioconda._

Brings in the latest mulled work that was migrated into galaxy-lib and made to work with combinations of requirements, local caching, non-bioconda channels, etc....

The local test can be run with the following commands:

```
virtualenv planemo-venv
. planemo-venv/bin/activate
pip install --upgrade pip 
pip install planemo 
planemo serve --galaxy_root . --mulled_containers test/functional/tools/mulled_example_multi_1.xml
# Open localhost:9090 and run the example tool - you will see Galaxy build the container on the fly.
```

Then open a web browser and run the new test tool - it should find the container you built and use it without any explicit markup required in the tool XML.

This demonstrates the local cached mulled containers - but remote ones on quay.io will also be used dynamically if `GALAXY_CONFIG_ENABLE_BETA_MULLED_CONTAINERS` is enabled - such as the ones @bgruening built for every package in bioconda.

Update: Replaced old manual test description with much more automated approach using new release of planemo leveraging all this magic. Old instructions were:

```
virtualenv .galaxy-lib-venv
. .galaxy-lib-venv/bin/activate
pip install galaxy-lib
mulled-build-tool build test/functional/tools/mulled_example_multi_1.xml  # Builds a mulled container locally for tool.
export GALAXY_CONFIG_ENABLE_BETA_MULLED_CONTAINERS=True
emacs config/job_conf.xml # Configure a Docker-enabled destination for mulled_example_multi_1
GALAXY_RUN_WITH_TEST_TOOLS=1 sh run.sh
```

### CODE REVIEW

File: test/functional/tools/mulled_example_multi_1.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<tool id="mulled_example_multi_1" name="mulled_example_multi_1" version="0.1.0">
+    <command><![CDATA[
```
Comment (bgruening) [inline test/functional/tools/mulled_example_multi_1.xml:2]: Should this tool test anything? It also fails `planemo lint` ;)

Comment (jmchilton) [reply test/functional/tools/mulled_example_multi_1.xml:2]: All the test tools fail planemo lint. As for the tests - I don't have any because this tool really depends on the dependency resolvers working which none of the other framework test tools do. I want them testing the core tool framework - not the ... lets say dependency resolution ecosystem.

**Comment by bgruening** [comment]
Obviously I'm a huge fan of this and because it contains code from us I will not vote. Just one thing this PR contains some useful `tools` for handling/searching mulled containers and I'm not sure these should be under `lib/`, but I see that these are useful in galaxy-lib for planemo etc...

Moreover, thanks a lot to @thriqon who's master thesis made all this possible!

File: lib/galaxy/tools/deps/container_resolvers/mulled.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                    continue
+                if not cached_image.package_name == target.package_name:
+                    continue
+                if not target.version or target.version == cached_image.version:
+                    image = cached_image
+                    break
+        else:
+            name = image_name(targets)
+            for cached_image in cached_images:
+                if name == cached_image.hash:
```
Comment (mvdbeek) [inline lib/galaxy/tools/deps/container_resolvers/mulled.py]: `cached_image` can be a`CachedMulledImageSingleTarget` that doesn't have a `hash` attribute, right? `if cached_image.multi_target and name == cached_image.hash:` should be OK

File: lib/galaxy/tools/deps/mulled/util.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    requests = None
+
+
+def quay_versions(namespace, pkg_name):
+    """Get all version tags for a Docker image stored on quay.io for supplied package name."""
+    if requests is None:
+        raise Exception("requets library is unavailable, functionality not available.")
+
+    url = 'https://quay.io/api/v1/repository/%s/%s' % (namespace, pkg_name)
+    response = requests.get(url, timeout=None)
```
Comment (mvdbeek) [inline lib/galaxy/tools/deps/mulled/util.py:23]: this fails with:

```
galaxy.tools.deps.mulled.util WARNING 2016-10-02 11:41:39,416 {u'status': 401, u'error_message': u'Requires authentication', u'title': u'invalid_token', u'error_type': u'invalid_token', u'detail': u'Requires authentication', u'type': u'https://quay.io/api/v1/error/invalid_token'}
```

Comment (mvdbeek) [reply lib/galaxy/tools/deps/mulled/util.py:23]: ~~fwiw, I'm using something like https://gist.github.com/mvdbeek/5e48d2c21c1440e22bda895769f5e450 for v2 registries.~~
seems that you get a `401` also if the repository doesn't exist ...

**Comment by mvdbeek** [comment]
Amazing! If there is only a single requirement to resolve it even works with my WIP udocker implementation.

**Comment by jmchilton** [comment]
Okay - I've pushed through another iteration on this. I've:
- Fixed the problem referenced by @mvdbeek if the mulled dependency is missing on quay.io.
- Added another container resolver that will actually just build the container as needed - it comes after checking if they are locally cached or available on quay.io.
- Made it so that involucro - the tool that handles the Docker interactions when building mulled containers - will install on-demand.
- Added more documentation to `galaxy.ini.sample`.
- Added the destination option `requires_container` - to force job destinations to run containerized jobs and fail if a relevant container is not configured (either by the tool author or deployer). This will enable planemo to force containers to be available during testing and quickly signal an obvious error if they are not.
- Update galaxy-lib with some other changes to enable all of this.

**Comment by jmchilton** [comment]
Did a release of planemo which again greatly reduces the barrier to testing this PR. The new `--mulled_containers` flag should take care of configuring the resolvers and setting a job_conf.xml with Docker enabled. I update the PR description with new testing instructions.

**Comment by natefoo** [comment]
:+1: this is amazing.

**Comment by bgruening** [comment]
Oh, we have a :+1:  @jmchilton we have a :+1:  :)
I think I need to adopt it to the new namespace, but hey we have a :+1: !!!!

**Comment by jmchilton** [comment]
Updated to include the latest fixes from @bgruening's work downstream in galaxy-lib (through version 16.10.7 of that library) and rebased into two clean, atomic commits.

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
to trigger retesting

**Review by martenson — APPROVED** [review]

**Comment by jmchilton** [comment]
Thanks for the reviews all and thanks for the merge @martenson!

**Comment by bgruening** [comment]
:tada: :balloon: I think I take a week off - this PR is awesome!
Docker for free for everyone!
